File: fuzzlab/web/static/js/downsample.js
// fuzzlab control panel — pure downsampling logic (U5, per R-05/R-08/R-09).
//
// Touches no `window`/`document` at import, so this module loads unmodified in
// both Node (`node --test`, R-09 layer A) and the browser (relative-path ESM,
// no build step). Two modes, both operating on COLUMNAR data `[xs, ys]` (uPlot's
// native shape; x in unix seconds):
//
//   - `lttb(xs, ys, threshold)` — Largest-Triangle-Three-Buckets. Preserves visual
//     shape (spikes/jumps) far better than naive bucket-averaging; used for scalar
//     trend lines (training curves, coverage growth, mutation reward/novelty).
//   - `envelope(xs, ys, threshold)` — per-bucket [min, max], for the bandit
//     posterior/regret CI band (R-08's "min/max envelope mode"): shows the true
//     spread within a bucket instead of erasing it the way an average would.
//
// Both keep the series's first and last point untouched and are no-ops when the
// series already has ≤ `threshold` points, so charts with little data never lose
// precision they didn't need to give up.

export function lttb(xs, ys, threshold) {
  const n = xs.length;
  if (threshold >= n || threshold <= 2 || n === 0) {
    return [xs.slice(), ys.slice()];
  }

  const outXs = [xs[0]];
  const outYs = [ys[0]];

  // Bucket size excludes the fixed first/last points.
  const bucketSize = (n - 2) / (threshold - 2);
  let a = 0; // index of the last picked point

  for (let i = 0; i < threshold - 2; i++) {
    // Average point of the NEXT bucket (for the triangle's third vertex).
    const nextStart = Math.floor((i + 1) * bucketSize) + 1;
    const nextEnd = Math.min(Math.floor((i + 2) * bucketSize) + 1, n);
    let avgX = 0, avgY = 0;
    const nextCount = Math.max(1, nextEnd - nextStart);
    for (let j = nextStart; j < nextEnd; j++) { avgX += xs[j]; avgY += ys[j]; }
    avgX /= nextCount;
    avgY /= nextCount;

    // This bucket's range to pick the largest triangle from.
    const rangeStart = Math.floor(i * bucketSize) + 1;
    const rangeEnd = Math.min(Math.floor((i + 1) * bucketSize) + 1, n);

    const ax = xs[a], ay = ys[a];
    let bestArea = -1, bestIdx = rangeStart;
    for (let j = rangeStart; j < rangeEnd; j++) {
      const area = Math.abs(
        (ax - avgX) * (ys[j] - ay) - (ax - xs[j]) * (avgY - ay)
      ) * 0.5;
      if (area > bestArea) { bestArea = area; bestIdx = j; }
    }
    outXs.push(xs[bestIdx]);
    outYs.push(ys[bestIdx]);
    a = bestIdx;
  }

  outXs.push(xs[n - 1]);
  outYs.push(ys[n - 1]);
  return [outXs, outYs];
}

export function envelope(xs, ys, threshold) {
  const n = xs.length;
  if (threshold >= n || threshold <= 0 || n === 0) {
    return [xs.slice(), ys.slice(), ys.slice()];
  }
  const outXs = [], outMin = [], outMax = [];
  const bucketSize = n / threshold;
  for (let i = 0; i < threshold; i++) {
    const start = Math.floor(i * bucketSize);
    const end = i === threshold - 1 ? n : Math.floor((i + 1) * bucketSize);
    if (start >= end) continue;
    let mn = ys[start], mx = ys[start], sumX = 0;
    for (let j = start; j < end; j++) {
      if (ys[j] < mn) mn = ys[j];
      if (ys[j] > mx) mx = ys[j];
      sumX += xs[j];
    }
    outXs.push(sumX / (end - start));
    outMin.push(mn);
    outMax.push(mx);
  }
  return [outXs, outMin, outMax];
}

File: fuzzlab/web/downsample.py
"""Server-side LTTB / envelope downsampling for diagnostics chart data (U5, R-05/R-08).

Per R-05's resolved render order ("downsample server-side -> send -> EMA
client-side"), a chart's `/api/diagnostics/series/data` payload is reduced to
at most ~1000 points *before* it goes over the wire — EMA smoothing then runs
client-side (`static/js/smoothing.js`) over the already-small series, so the
smoothing slider is instant and never re-fetches.

This is a straight, dependency-free Python port of the same two algorithms
implemented for the browser in `static/js/downsample.js` (LTTB for trend
lines; per-bucket min/max envelope for the bandit posterior/regret CI band).
Keeping both is deliberate, not duplication for its own sake: the browser
copy is unit-tested via `node --test` (R-09 layer A) and doubles as the logic
a future client-side re-downsample-on-zoom would use, while this copy is what
actually shrinks the JSON response.
"""

from __future__ import annotations


def lttb(xs: list[float], ys: list[float], threshold: int) -> tuple[list[float], list[float]]:
    """Largest-Triangle-Three-Buckets downsampling. Keeps the first/last point;
    a no-op when the series already has ``threshold`` points or fewer."""
    n = len(xs)
    if threshold >= n or threshold <= 2 or n == 0:
        return list(xs), list(ys)

    out_xs = [xs[0]]
    out_ys = [ys[0]]
    bucket_size = (n - 2) / (threshold - 2)
    a = 0

    for i in range(threshold - 2):
        next_start = int((i + 1) * bucket_size) + 1
        next_end = min(int((i + 2) * bucket_size) + 1, n)
        next_count = max(1, next_end - next_start)
        avg_x = sum(xs[next_start:next_end]) / next_count
        avg_y = sum(ys[next_start:next_end]) / next_count

        range_start = int(i * bucket_size) + 1
        range_end = min(int((i + 1) * bucket_size) + 1, n)

        ax, ay = xs[a], ys[a]
        best_area, best_idx = -1.0, range_start
        for j in range(range_start, range_end):
            area = abs((ax - avg_x) * (ys[j] - ay) - (ax - xs[j]) * (avg_y - ay)) * 0.5
            if area > best_area:
                best_area, best_idx = area, j
        out_xs.append(xs[best_idx])
        out_ys.append(ys[best_idx])
        a = best_idx

    out_xs.append(xs[n - 1])
    out_ys.append(ys[n - 1])
    return out_xs, out_ys


def envelope(xs: list[float], ys: list[float],
             threshold: int) -> tuple[list[float], list[float], list[float]]:
    """Per-bucket (x, min(y), max(y)) — the CI-band mode for the bandit charts
    (R-08). A no-op when the series already has ``threshold`` points or fewer."""
    n = len(xs)
    if threshold >= n or threshold <= 0 or n == 0:
        return list(xs), list(ys), list(ys)

    out_xs: list[float] = []
    out_min: list[float] = []
    out_max: list[float] = []
    bucket_size = n / threshold
    for i in range(threshold):
        start = int(i * bucket_size)
        end = n if i == threshold - 1 else int((i + 1) * bucket_size)
        if start >= end:
            continue
        chunk = ys[start:end]
        out_xs.append(sum(xs[start:end]) / (end - start))
        out_min.append(min(chunk))
        out_max.append(max(chunk))
    return out_xs, out_min, out_max
